core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java

```diff
@@ -75,14 +86,105 @@ public DnsResolverGroupBuilder resolvedAddressTypes(
         return this;
     }
 
+    /**
+     * Sets whether to enable auto refresh for expired {@link DnsRecord}s.
+     * This option is enabled by default.
+     *
+     * <p>If disable this option, the expired {@link DnsRecord} is removed from the {@link DnsCache} and a new
+     * {@link DnsQuery} will be executed when a connection is newly established.
+     * The DNS time may make the total of duration of a request longer. So it is recommended to turn on this
+     * option when a <strong>limited</strong> number of hostnames are used.
+     */
+    @UnstableApi
+    public DnsResolverGroupBuilder enableAutoRefresh(boolean autoRefresh) {
+        this.autoRefresh = autoRefresh;
+        return this;
+    }
+
+    /**
+     * Sets the {@link Backoff} which is used when the {@link DnsNameResolver} fails to update the cache.
+     */
+    public DnsResolverGroupBuilder autoRefreshBackoff(Backoff refreshBackoff) {
+        autoRefreshBackoff = requireNonNull(refreshBackoff, "refreshBackoff");
+        return this;
+    }
+
+    /**
+     * Sets the {@link ToLongFunction} which determines how long the {@link DnsRecord}s for a hostname should
+     * be refreshed after the first cache.
+     *
+     * <p>For example:
+     * <pre>{@code
+     * dnsResolverGroupBuilder.autoRefreshTimeout(hostname -> {
+     *     if (hostname.endsWith("busy.domain.com")) {
+     *         return Duration.ofDays(7).toMillis(); // Automatically refresh the cached domain for 7 days.
+     *     }
+     *     if (hostname.endsWith("sporadic.domain.dom")) {
+     *         return 0; // Don't need to refresh a sporadically used domain.
+     *     }
+     *     ...
+     * });
+     * }</pre>
+     *
+     * <p>Note this method is mutually exclusive with {@link #autoRefreshTimeout(Duration)} and
+     * {@link #autoRefreshTimeoutMillis(long)}.
+     */
+    @UnstableApi
+    public DnsResolverGroupBuilder autoRefreshTimeout(ToLongFunction<? super String> timeoutFunction) {
```

Comment:
IIRC, we always allow to call setter methods multiple times on a builder:
```java
DnsResolverGroupBuilder builder = ..;
val first = builder.autoRefreshTimeout(...).build();
val second = builder.autoRefreshTimeout(...).build();
```
How about fixing it?

Reply:
I blocked that because `autoRefreshTimeoutMillis()` also calls this method.
I wanted to prevent the following case.
```
builder.autoRefreshTimeout(Duration.ofDays(1))
       .autoRefreshTimeout(hostname -> ...);  // Duration.ofDays
       .build();
```
Additionally, users can not create the `DnsResolverGroupBuilder` which can be accessed by the `ClientFactoryBuilder.domainNameResolverCustomizer`.

Reply:
We allow to override the properties if they are set by the same method name.
For example https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java#L474-L492
Should we also fix them? What's the rational behavior you think?

Reply:
I agree that we allow overriding properties of a builder.
It should be useful if a builder object is reused to create new objects. It is worth supporting overriding.

The constructor of `DnsResolverGroupBuilder` is not exposed externally and it is used only once to create an `AddressResolverGroup`,
I don't see users need to call two different timeout configurations for a `DnsResolverGroupBuilder`.

I want to enforce users to carefully use the options, but agree that there is no serious harm to allow overriding.
If you still think allowing overriding is more sensible behavior, I'm willing to address it. :-)

Reply:
That makes sense. Let's keep as is. 👍

Reply:
Final thought is commented on https://github.com/line/armeria/pull/4400#discussion_r964444919